lib/matplotlib/dates.py

```diff
@@ -739,12 +739,239 @@ def __call__(self, x, pos=0):
         return num2date(self.t[ind], self.tz).strftime(self.fmt)
 
 
+class ConciseDateFormatter(ticker.Formatter):
+    """
+    This class attempts to figure out the best format to use for the
+    date, and to make it as compact as possible, but still be complete. This is
+    most useful when used with the :class:`AutoDateLocator`.
+
+    >>> locator = AutoDateLocator()
+    >>> formatter = ConciseDateFormatter(locator)
+
+    Parameters
+    ----------
+
+    locator : `.ticker.Locator`
+        Locator that this axis is using.
+
+    tz : string, optional
+        Passed to `.dates.date2num`.
+
+    formats : list of 6 strings, optional
+        Format strings for 6 levels of tick labelling: mostly years,
+        months, days, hours, minutes, and seconds.  Strings use
+        the same format codes as `strftime`.  Default is
+        ``['%Y', '%b', '%d', '%H:%M', '%H:%M', '%S.%f'] ``
+
+    zeroformats : list of 6 strings, optional
+        Format strings for tick labels that are "zeros" for a given tick
+        level.  i.e. if most ticks are months, ticks around 1 Jan 2005 will be
+        labeled "Dec", "2005", "Feb".  The default is
+        ``['', '%Y', '%b', '%b-%d', '%H:%M', '%H:%M']``
+
+    offsetformats : list of 6 strings, optional
+        Format string for the seven levels that is applied to the "offset"
+        string found on the right side of an x-axis, or top of a y-axis.
+        Combined with the tick labels this should completely specify the
+        date.  The default is::
+
+            ['', '%Y', '%Y-%b', '%Y-%b-%d', '%Y-%b-%d', '%Y-%b-%d %H:%M']
+
+    show_offset : bool
+        Whether to show the offset or not.  Default is ``True``.
+
+    Examples
+    --------
+
+    See :doc:`gallery/ticks_and_spines/date_concise_formater`
+
+    .. plot::
+
+        base = datetime.datetime(2005, 2, 1)
+        dates = np.array([base + datetime.timedelta(hours=(2 * i))
+                          for i in range(732)])
+        N = len(dates)
+        np.random.seed(19680801)
+        y = np.cumsum(np.random.randn(N))
+
+        fig, axs = plt.subplots(3, 1, constrained_layout=True)
+        for nn, ax in enumerate(axs[0]):
+            locator = mdates.AutoDateLocator()
+            formatter = mdates.ConciseDateFormatter(locator)
+            ax.xaxis.set_major_locator(locator)
+            ax.xaxis.set_major_formatter(formatter)
+
+            ax.plot(dates, y)
+            ax.set_xlim(lims[nn])
+        axs[0].set_title('Concise Date Formatter')
+
+    """
+
+    def __init__(self, locator, tz=None, formats=None, offsetformats=None,
```

Comment:
offset_formats / zero_formats? (consistent with show_offset)

Reply:
also tempted to just put the defaults in the signature `formats=('%Y', ...), ...`; not sure supporting None is really necessary

Reply:
That kinda makes the signature ugly though, and we still need `None` for zero formats anyway...